File: SampleAppObjC/PPHSDKSampleApp/OfflineModeViewController.m
//
//  OfflineModeViewController.m
//  PPHSDKSampleApp
//
//  Created by Patil, Mihir on 7/5/18.
//  Copyright © 2018 Patil, Mihir. All rights reserved.
//

#import "OfflineModeViewController.h"
#import "UIButton+CustomButton.h"

@interface OfflineModeViewController ()
@property (weak, nonatomic) IBOutlet UISwitch *offlineModeSwitch;
@property (weak, nonatomic) IBOutlet UIButton *getOfflineStatusBtn;
@property (weak, nonatomic) IBOutlet UITextView *getOfflineStatusCodeTxtView;
@property (weak, nonatomic) IBOutlet UIButton *replayOfflineTransactionBtn;
@property (weak, nonatomic) IBOutlet UITextView *replayOfflineTransactionCodeTxtView;
@property (weak, nonatomic) IBOutlet UIButton *stopReplayBtn;
@property (weak, nonatomic) IBOutlet UITextView *stopReplayCodeTxtView;
@property (weak, nonatomic) IBOutlet UITextView *replayTransactionResultsTextView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *replayTransactionIndicatorView;
@property (weak, nonatomic) IBOutlet UILabel *offlineModeLabel;
@end

@implementation OfflineModeViewController

BOOL offlineInit = NO;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setUpDefaultView];
}

- (void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
    [_offlineDelegate offlineModeController: self offline: self.offlineMode];
}

// If the offlineModeSwitch is toggled. Set the value for the offlineMode Flag which will make the appropriate call
// to the SDK.
// - Parameter sender: UISwitch for the offlineMode.
- (IBAction)offlineModeSwitchPressed:(UISwitch *)sender {
    [self toggleOfflineMode];
}

// If offlineMode is set to true then we will start taking offline payments and if it is set to false
// then we stop taking offline payments. To Start/Stop taking offline payments, we ned to make a call to
// the SDK. If we start taking online payments then we MUST call the stopOfflinePayment() in order to start
// taking live payments again.
-(void) toggleOfflineMode {
    if(!self.offlineMode) {
        if ([[PayPalRetailSDK transactionManager] getOfflinePaymentEligibility]) {
            [[PayPalRetailSDK transactionManager] startOfflinePayment:^(PPRetailError *error, NSArray *status) {
                if (error != nil){
                    NSLog(@"%@", error.developerMessage);
                } else {
                    // Check to see if OfflinePayment is enabled
                    if ([[PayPalRetailSDK transactionManager] getOfflinePaymentEnabled]){
                        [self updateOfflineModeUI];
                    }
                    [self offlineTransactionStatusList:status];
                }
            }];
        } else {
            NSLog(@"Merchant is not eligible to take Offline Payments");
        }
    } else {
        [[PayPalRetailSDK transactionManager] stopOfflinePayment:^(PPRetailError *error, NSArray *status) {
            if (error != nil){
                NSLog(@"Error: %@", error.debugDescription);
            } else {
                [self offlineTransactionStatusList:status];
            }
        }];
        [self updateOfflineModeUI];
    }
}

// The function will get the offline Status. It is a callback. It will give you an array of status which will
// tell you about the status of the payment.
// - Parameter sender: UIButton assoicated with the Get Offline Status button.
- (IBAction)getOfflineStatus:(UIButton *)sender {
    [[PayPalRetailSDK transactionManager] getOfflinePaymentStatus:^(PPRetailError *error, NSArray *status) {
        if(error != nil) {
            NSLog(@"Error: %@", error.description);
        } else {
            [self offlineTransactionStatusList:status];
        }
    }];
}

// If payments are taken in offline mode then those payments are saved on the device. This function, if the
// device is online, will go through those payments saved on the device and process those payments.
// The call back will give you the result whether those payments are completed, failed or were declined.
// - Parameter sender: UIButton associated with "Replay Offline Transaction" button
- (IBAction)replayOfflineTransaction:(UIButton *)sender {
    
    
    if (offlineInit){
        [self showReplayTransactionAlertForOfflineInit:YES];
    } else {
        if (_offlineMode){
           [self showReplayTransactionAlertForOfflineInit:NO];
        }
        
        [self replayTransactionAnimation:YES];
        [[PayPalRetailSDK transactionManager] startReplayOfflineTxns:^(PPRetailError *error, NSArray *status) {
            [self updateOfflineModeUI];
            [self replayTransactionAnimation:YES];
            if(error != nil) {
                NSLog(@"Error: %@", error.description);
            } else {
                [self offlineTransactionStatusList:status];
            }
        }];
    }
}

// If we are replaying transactions and we want to stop replayingTransactions then we can call this function.
// For example: If you went offline when replaying transactions.
// - Parameter sender: UIButton associated with "Stop Replay" Button
- (IBAction)stopReplay:(UIButton *)sender {
    [self.replayTransactionIndicatorView stopAnimating];
    [[PayPalRetailSDK transactionManager] stopReplayOfflineTxns:^(PPRetailError *error, NSArray *status) {
        if (error != nil) {
            NSLog(@"Stopped replaying offline transactions");
        } else {
            [self offlineTransactionStatusList:status];
        }
    }];
}

-(void) offlineTransactionStatusList:(NSArray *)status{
    int uncompleted = 0;
    int completed = 0;
    int failed = 0;
    int declined = 0;
    for (PPRetailOfflinePaymentStatus *s in status) {
        if(s.errNo == 0) {
            if(s.retry > 0) {
                completed++;
            } else {
                uncompleted++;
            }
        } else if(s.isDeclined) {
            declined++;
        } else {
            failed++;
        }
    }
    self.replayTransactionResultsTextView.text = [NSString stringWithFormat:@"Results:\n Uncompleted: %d \n Completed: %d \n Failed: %d \n Declined: %d",uncompleted,completed,failed,declined];
    self.stopReplayBtn.enabled = NO;
}

-(void)setDelegate:(UIViewController *)delegateController{
    _offlineDelegate = delegateController;
}

-(void)setUpDefaultView{
    // Set the offlineMode switch on/off according to the value passed from PaymentViewController. Originally false.
    [self.offlineModeSwitch setOn:self.offlineMode];
    [self updateOfflineModeUI];
    // Stop Replay Button is only needed when we are replaying transactions. Otherwise it is disabled.
    self.stopReplayBtn.enabled = NO;
    self.getOfflineStatusCodeTxtView.text = @"[[PayPalRetailSDK transactionManager]  getOfflinePaymentStatus:^(PPRetailError *error, NSArray *status) {\n <code to handle success/failure> \n}];";
    self.replayOfflineTransactionCodeTxtView.text = @"[[PayPalRetailSDK transactionManager]  startReplayOfflineTxns:^(PPRetailError *error, NSArray *status) {\n <code to handle success/failure> \n}];";
    self.stopReplayCodeTxtView.text = @"[[PayPalRetailSDK transactionManager] stopReplayOfflineTxns];";
    self.replayTransactionResultsTextView.text = @"";
    [self offlineSDKInit];
    [self updateOfflineModeUI];
    
    [CustomButton customizeButton:_getOfflineStatusBtn];
    [CustomButton customizeButton:_replayOfflineTransactionBtn];
    [CustomButton customizeButton:_stopReplayBtn];
}

-(void) offlineSDKInit{
    NSUserDefaults *tokenDefault =  [NSUserDefaults standardUserDefaults];
    offlineInit = [tokenDefault boolForKey:@"offlineSDKInit"];
    if (offlineInit){
        self.offlineModeSwitch.enabled = NO;
    }
}

// THIS FUNCTION IS ONLY FOR UI. This function will enable/disable "Replay Transaction" Button
// depending on if the offlineMode is on or off.
-(void)updateOfflineModeUI{
    if ([[PayPalRetailSDK transactionManager] getOfflinePaymentEnabled]){
        _offlineMode = YES;
        _offlineModeLabel.text = @"ENABLED";
        _offlineModeLabel.textColor = UIColor.greenColor;
        [_offlineModeSwitch setOn:YES];
    } else {
        _offlineMode = NO;
        _offlineModeLabel.text = @"";
        _offlineModeLabel.textColor = UIColor.redColor;
        [_offlineModeSwitch setOn:NO];
    }
}

-(void) replayTransactionAnimation:(BOOL) start {
    if (start){
        [self.replayTransactionIndicatorView startAnimating];
        [self.replayOfflineTransactionBtn setHidden:YES];
        self.stopReplayBtn.enabled = YES;
    } else {
        [self.replayTransactionIndicatorView stopAnimating];
        [self.replayOfflineTransactionBtn setHidden:NO];
        self.stopReplayBtn.enabled = NO;
    }
}

-(void) showReplayTransactionAlertForOfflineInit:(BOOL) offlineInit {
    NSString *title = @"Replaying while in Offline Mode";
    NSString *message = @"Replaying transaction in offlineMode will bring the SDK back into Online Mode";
    if (offlineInit){
        title = @"Cannot Replay in Offline Init";
        message = @"Replay is not allowed while the SDK is initialized in offline Mode.";
    }
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: SampleApp/PPHSDKSampleApp/OfflineModeViewController.swift
//
//  OfflineModeViewController.swift
//  PPHSDKSampleApp
//
//  Created by Deol, Sukhpreet(AWF) on 6/21/18.
//  Copyright © 2018 cowright. All rights reserved.
//

import UIKit
import PayPalRetailSDK

protocol OfflineModeViewControllerDelegate: NSObjectProtocol {
    func offlineMode(controller: OfflineModeViewController, didChange isOffline: Bool)
}

class OfflineModeViewController: UIViewController {
    
    @IBOutlet weak var offlineModeSwitch: UISwitch!
    @IBOutlet weak var getOfflineStatusBtn: CustomButton!
    @IBOutlet weak var getOfflineStatusCodeTxtView: UITextView!
    @IBOutlet weak var replayOfflineTransactionBtn: CustomButton!
    @IBOutlet weak var replayOfflineTransactionCodeTxtView: UITextView!
    @IBOutlet weak var replayTransactionIndicatorView: UIActivityIndicatorView!
    @IBOutlet weak var stopReplayBtn: CustomButton!
    @IBOutlet weak var stopReplayCodeTxtView: UITextView!
    @IBOutlet weak var replayTransactionResultsTextView: UITextView!
    @IBOutlet weak var offlineModeLabel: UILabel!
    
    // Local Flag for offline Mode
    var offlineMode: Bool!
    var offlineSDK: Bool = false
    
    weak var delegate: OfflineModeViewControllerDelegate?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setUpDefaultView()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.delegate?.offlineMode(controller: self, didChange: self.offlineMode)
    }
    
    /// If the offlineModeSwitch is toggled. Call the toggleOfflineMode() function.
    /// - Parameter sender: UISwitch for the offlineMode.
    @IBAction func offlineModeSwitchPressed(_ sender: UISwitch) {
        toggleOfflineMode()
    }
    
    /// If offlineMode is set to true then we will start taking offline payments and if it is set to false
    /// then we stop taking offline payments. To Start/Stop taking offline payments, we ned to make a call to
    /// the SDK. If we start taking online payments then we MUST call the stopOfflinePayment() in order to start
    /// taking live payments again.
    /// FOR UI ONLY - Change the UI based on offlineMode flag.
    private func toggleOfflineMode(){
        if !offlineMode {
            // If the merchant is whitelisted to take offline Payments.
            if (PayPalRetailSDK.transactionManager()?.getOfflinePaymentEligibility())! {
                // This call with initalize offlinePayment. Any kind of failure will be returned to this callback.
                // Along with the status of the transaction that were done during offline Mode.
                PayPalRetailSDK.transactionManager()?.startOfflinePayment({ (error, info) in
                    if error != nil {
                        print(error?.developerMessage ?? "There was a problem initializing offlinePayment.")
                    } else {
                        // Check to see if OfflinePayment was enabled
                        if (PayPalRetailSDK.transactionManager()?.getOfflinePaymentEnabled())! {
                            self.updateOfflineModeUI()
                        }
                        self.offlineTransactionStatusList(statusList: info?.statusList)
                    }
                })
            } else {
                print("Merchant is not eligible to take Offline Payments")
                self.alertMerchantNotEligableForOfflinePayments()

            }
        } else {
            // Turn off offlineMode
            PayPalRetailSDK.transactionManager()?.stopOfflinePayment({ (error, info) in
                if error != nil {
                    print("Error: \(error.debugDescription)")
                } else {
                    self.offlineTransactionStatusList(statusList: info?.statusList)
                }
            })
            self.updateOfflineModeUI()
        }
    }
    
    /// The function will get the offline Status. It is a callback. It will give you an array of status which will
    /// tell you about the status of the payment.
    /// - Parameter sender: CustomButton assoicated with the Get Offline Status button.
    @IBAction func getOfflineStatus(_ sender: CustomButton) {
        PayPalRetailSDK.transactionManager().getOfflinePaymentStatus { (error, info) in
            if error  != nil {
                print("Error: \(error.debugDescription)")
            } else {
                self.offlineTransactionStatusList(statusList: info?.statusList)
            }
        }
    }
    
    /// If payments are taken in offline mode then those payments are saved on the device. This function, if the
    /// device is online, will go through those payments saved on the device and process those payments.
    /// The call back will give you the result whether those payments are completed, failed or were declined.
    /// - Parameter sender: CustomButton associated with "Replay Offline Transaction" button
    @IBAction func replayOfflineTransaction(_ sender: CustomButton) {
        
        if offlineSDK {
            showReplayTransactionAlert(offlineInit: true)
        } else {
            if offlineMode {
                showReplayTransactionAlert(offlineInit: false)
            }
            
            self.replayTransactionAnimation(true)
            PayPalRetailSDK.transactionManager().startReplayOfflineTxns { [unowned self] (error, info) in
                self.updateOfflineModeUI()
                self.replayTransactionAnimation(false)
                if error != nil {
                    print("Error is: ", error.debugDescription)
                } else {
                    self.offlineTransactionStatusList(statusList: info?.statusList)
                }
            }
        }
    }
    
    /// If we are replaying transactions and we want to stop replayingTransactions then we can call this function.
    /// For example: If you went offline when replaying transactions.
    /// - Parameter sender: CustomButton associated with "Stop Replay" Button
    @IBAction func stopReplay(_ sender: CustomButton) {
        replayTransactionIndicatorView.stopAnimating()
        replayOfflineTransactionBtn.isHidden = false
        PayPalRetailSDK.transactionManager()?.stopReplayOfflineTxns({ (error, info) in
            if error != nil {
                print("Stopped replaying offline transactions")
            } else {
                self.offlineTransactionStatusList(statusList: info?.statusList)
            }
        })
    }
    
    private func offlineTransactionStatusList(statusList: [Any]?){
        guard let statusArray: [PPRetailOfflinePaymentStatus] = statusList as? [PPRetailOfflinePaymentStatus] else {return}
        var uncompleted: Int = 0
        var completed: Int = 0
        var failed: Int = 0
        var declined: Int = 0
        
        for status in statusArray {
            if status.errNo == 0 {
                if status.retry > 0 {
                    completed += 1
                } else {
                    uncompleted += 1
                }
            } else if status.isDeclined {
                declined += 1
            } else {
                failed += 1
            }
        }
        self.replayTransactionResultsTextView.text = "Uncompleted: \(uncompleted) \nCompleted: \(completed) \nFailed: \(failed) \nDeclined: \(declined)"
        self.stopReplayBtn.isEnabled = false
    }
    
    private func setUpDefaultView(){
        /// Set the offlineMode switch on/off according to the value passed from PaymentViewController. Originally false.
        offlineModeSwitch.isOn = offlineMode
        // Stop Replay Button is only needed when we are replaying transactions. Otherwise it is disabled.
        stopReplayBtn.isEnabled = false
        getOfflineStatusCodeTxtView.text = "PayPalRetailSDK.transactionManager().getOfflinePaymentStatus({ (error, statusList) in // Code })"
        replayOfflineTransactionCodeTxtView.text = "PayPalRetailSDK.transactionManager().startReplayOfflineTxns({ (error, statusList) in // Code })"
        stopReplayCodeTxtView.text = "PayPalRetailSDK.transactionManager().stopReplayOfflineTxns()"
        self.replayTransactionResultsTextView.text = "Uncompleted: 0 \nCompleted: 0 \nFailed: 0 \nDeclined: 0"
        updateOfflineModeUI()
        self.navigationItem.backBarButtonItem = UIBarButtonItem(title: "", style: .plain, target: nil, action: nil)
        offlineSDKInit()
    }
    
    @objc private func offlineSDKInit(){
        let tokenDefault = UserDefaults.init()
        guard (tokenDefault.value(forKey: "offlineSDKInit") != nil) else { return }
        self.offlineSDK = tokenDefault.bool(forKey: "offlineSDKInit")
        self.offlineModeSwitch.isEnabled = false
    }
    
    private func updateOfflineModeUI(){
        if (PayPalRetailSDK.transactionManager()?.getOfflinePaymentEnabled())! {
            offlineMode = true
            offlineModeLabel.text = "ENABLED"
            offlineModeLabel.textColor = .green
            offlineModeSwitch.isOn = true
        } else {
            offlineMode = false
            offlineModeLabel.text = ""
            offlineModeLabel.textColor = .red
            offlineModeSwitch.isOn = false
        }
    }
    
    private func replayTransactionAnimation(_ start: Bool){
        if start {
            replayTransactionIndicatorView.startAnimating()
            replayOfflineTransactionBtn.isHidden = true
            stopReplayBtn.isEnabled = true
        } else {
            replayTransactionIndicatorView.stopAnimating()
            replayOfflineTransactionBtn.isHidden = false
            stopReplayBtn.isEnabled = false
        }
    }
    
    private func showReplayTransactionAlert(offlineInit: Bool){
        var title: String = "Replaying while in Offline Mode"
        var message: String = "Replaying transaction in offlineMode will bring the SDK back into Online Mode"
        
        if offlineInit {
            title = "Cannot Replay in Offline Init"
            message = "Replay is not allowed while the SDK is initialized in offline Mode."
        }
        
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let okAction = UIAlertAction(title: "OK", style: .default, handler: nil)
        alert.addAction(okAction)
        self.present(alert, animated: true, completion: nil)
    }
    
    fileprivate func alertMerchantNotEligableForOfflinePayments() {
        let alertController = UIAlertController(title: "Whoops!", message: "Merchant is not eligible to take Offline Payments", preferredStyle: UIAlertControllerStyle.alert)
        
        let okAction = UIAlertAction(title: "OK", style: .default) { _ in
            self.offlineModeSwitch.setOn(false, animated: true)
        }
        
        alertController.addAction(okAction)
        self.present(alertController, animated: true, completion: nil)
    }
}
